faunadb-scala/src/main/scala/faunadb/streaming/BodyValueFlowProcessor.scala

```diff
@@ -0,0 +1,92 @@
+
+import com.fasterxml.jackson.databind.{JsonNode, ObjectMapper}
+import faunadb.{QueryError, QueryErrorResponse}
+import faunadb.errors.{PermissionDeniedException, UnknownException}
+import faunadb.values.{StringV, VSuccess, Value}
+import org.slf4j.LoggerFactory
+
+import scala.collection.JavaConverters.asScalaIteratorConverter
+import scala.util.Try
+
+private [faunadb] class BodyValueFlowProcessor(json: ObjectMapper, syncLastTxnTime: Long => Unit) extends SubmissionPublisher[Value] with Flow.Processor[util.List[ByteBuffer], Value] {
+  private val log = LoggerFactory.getLogger(getClass)
+  private var subscription: Flow.Subscription = _
+  private var subscriber: Flow.Subscriber[_ >: Value] = _
+
+  // We do not request data from the publisher until we have one subscriber
+  override def subscribe(subscriber: Flow.Subscriber[_ >: Value]): Unit = {
+    if (this.subscriber == null) {
+      this.subscriber = subscriber
+      super.subscribe(subscriber)
+      requestOne()
+    } else
+      throw new IllegalStateException("BodyValueFlowProcessor can have only one subscriber")
+  }
+
+  override def onSubscribe(subscription: Flow.Subscription): Unit =
+    this.subscription = subscription
+
+  override def onNext(item: util.List[ByteBuffer]): Unit = {
+    import java.nio.charset.StandardCharsets
+    val text = item.iterator().asScala.map(StandardCharsets.UTF_8.decode(_).toString).mkString("")
+    Try {
+      val eventJsonNode = json.readTree(text)
+      val value = json.treeToValue[Value](eventJsonNode, classOf[Value])
+
+      // syncLastTxnTime if possible
+      value("txn").to[Long].toOpt.foreach(syncLastTxnTime)
+
+      // handle error in stream
+      isUnrecoverableError(value, eventJsonNode) match {
+        case None => submit(value)
+        case Some(unrecoverableError) =>
+          subscriber.onError(unrecoverableError) // notify subscriber stream
+          subscription.cancel() // cancel subscription on the request body
+      }
+    }.recover {
+      case e: Throwable =>
+        log.error(s"could not handle event $text", e)
+        subscriber.onError(e) // notify subscriber stream
+        subscription.cancel() // cancel subscription on the request body
+    }
+
+    requestOne()
+  }
+
+  private def isUnrecoverableError(event: Value, jsonNode: JsonNode): Option[Throwable] = {
+    (event("type"), Option(jsonNode.get("event"))) match {
+      case (VSuccess(StringV("error"), _), Some(error)) =>
+        val queryError = json.treeToValue(error, classOf[QueryError])
+        if (queryError.code == "permission denied") {
+          val queryErrorResponse = QueryErrorResponse(401, Vector(queryError))
+          Some(new PermissionDeniedException(queryErrorResponse))
+        } else {
+          // we are not expecting other kind of errors but we are reporting it
+          val queryErrorResponse = QueryErrorResponse(500, Vector(queryError))
+          Some(new UnknownException(queryErrorResponse))
+        }
```

Comment:
There's a new PR in core that will return both "time out" and "request timeout" errors. That'll force us to adhere to a more generic error handing approach here.

Reply:
Perhaps we can simplify this by creating a generic stream error that wraps the query error returned?

Reply:
I like the idea, I am proposing the following https://github.com/fauna/faunadb-jvm/pull/272/commits/1ee49b920a0a127f433055a54547010ec6e00466

WDYT?

Reply:
That's good.

Reply:
I'd prefer reusing the already existing Exceptions if possible here (e.g., mapping errors to `NotFoundException`, `PermissionDeniedException`, etc). From a user perspective, it will provide a better development experience. Users are most probably relying on the Exception types (rather than the Exception msg or the `QueryErrorResponse`) for mapping the errors coming from the driver to something meaningful to their domains. With the approach we are taking here (i.e., the generic `StreamingException`), the only thing they have to map/interpret the error is the Exception msg (which might not be very handy)...

Since it seems some further work regarding errors is coming down the pipe from core, I'd say we keep this bit of code here as is for this first iteration. This is something we should definitely improve later on though.

@agourlay, would you mind adding a `// TODO` comment here? :)

```scala
// TODO: map query errors to any of the existing 
// faunadb.errors.FaunaException
val ex = new StreamingException(queryError)
```

Thank you!